core/domain/blog_services.py

```diff
@@ -0,0 +1,613 @@
+    """
+    return blog_models.BlogPostModel.generate_new_blog_post_id()
+
+
+def get_blog_post_rights_from_model(blog_post_rights_model):
+    """Returns a blog post rights domain object given a blog post rights
+    model loaded from the datastore.
+
+    Args:
+        blog_post_rights_model: BlogPostRightsModel. The blog post rights model
+            loaded from the datastore.
+
+    Returns:
+        BlogPostRights. A blog post rights domain object corresponding to the
+        given blog post rights model.
+    """
+
+    return blog_domain.BlogPostRights(
+        blog_post_rights_model.id,
+        blog_post_rights_model.editor_ids,
+        blog_post_rights_model.blog_post_is_published)
+
+
+def get_blog_post_rights(blog_post_id, strict=True):
+    """Retrieves the rights object for the given blog post.
+
+    Args:
+        blog_post_id: str. ID of the blog post.
+        strict: bool. Whether to fail noisily if no blog post rights model
+            with a given ID exists in the datastore.
+
+    Returns:
+        BlogPostRights. The rights object associated with the given blog post.
+
+    Raises:
+        EntityNotFoundError. The blog post with ID blog post id was not
+            found in the datastore.
+    """
+
+    model = blog_models.BlogPostRightsModel.get(blog_post_id, strict=strict)
+
+    if model is None:
+        return None
+
+    return get_blog_post_rights_from_model(model)
+
+
+def get_blog_post_rights_by_user_id(user_id):
+    """Retrieves the rights object for all blog posts for which the given
+    user is an editor.
+
+    Args:
+        user_id: str. ID of the user.
+
+    Returns:
+        list(BlogPostRights). The rights objects associated with the
+        blog posts assigned to given user.
+    """
+    blog_post_rights_models = (
+        blog_models.BlogPostRightsModel.get_by_user(user_id))
+    return [
+        get_blog_post_rights_from_model(model)
+        for model in blog_post_rights_models
+        if model is not None]
+
+
+def does_blog_post_with_url_fragment_exist(url_fragment):
+    """Checks if blog post with provided url fragment exists.
+
+    Args:
+        url_fragment: str. The url fragment for the blog post.
+
+    Returns:
+        bool. Whether the the url fragment for the blog post exists.
+
+    Raises:
+        Exception. Blog Post URL fragment is not a string.
+    """
+    if not isinstance(url_fragment, python_utils.BASESTRING):
+        raise utils.ValidationError(
+            'Blog Post URL fragment should be a string. Recieved:'
+            '%s' % url_fragment)
+    existing_blog_post = get_blog_post_by_url_fragment(url_fragment)
+    return existing_blog_post is not None
+
+
+def _save_blog_post(blog_post):
+    """Saves a BlogPost domain object to the datastore.
+
+    Args:
+        blog_post: BlogPost. The blog post domain object for the given
+            blog post.
+    """
+    model = blog_models.BlogPostModel.get(blog_post.id, strict=True)
+    blog_post.validate()
```

Comment:
Should we validate based on the published status?

Reply:
Also, should we internally do a strict validation if the blog is published instead of service layer deciding when to do strict validation? (not sure!)

Reply:
We only have 2 ways to save a blog post, either as a draft or as published. We don't have any button which allows saving the post without clicking on any one of these.  Whenever a blog post is published strict validation is performed.(i.e even if its already published and user makes changes, the blog post will again have to be published)

Reply:
>  We don't have any button which allows saving the post without clicking on any one of these.

You can assume backend APIs to be independent i.e, one can hit URL without the frontend.

I'm fine with not internally maintaining the strict validation but I think the _save_blog_post method should validate strictly based on the publication status.